test/StreamJsonRpc.Tests/JsonRpcDelegatedDispatchTests.cs

```diff
@@ -0,0 +1,148 @@
+        var streams = Nerdbank.FullDuplexStream.CreateStreams();
+
+        this.clientRpc = new DelegatedJsonRpc(new HeaderDelimitedMessageHandler(streams.Item1));
+        this.serverRpc = new DelegatedJsonRpc(new HeaderDelimitedMessageHandler(streams.Item2), this.server);
+
+        this.serverRpc.TraceSource = new TraceSource("Server", SourceLevels.Information);
+        this.clientRpc.TraceSource = new TraceSource("Client", SourceLevels.Information);
+
+        this.serverRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+        this.clientRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+
+        this.clientRpc.StartListening();
+        this.serverRpc.StartListening();
+    }
+
+    [Fact]
+    public async Task DispatchRequestIsPassedCorrectTypeOfRequest()
+    {
+        await this.clientRpc.InvokeAsync<string>(nameof(Server.TestMethodAsync));
+        Assert.Equal("StreamJsonRpc.JsonMessageFormatter+JsonRpcRequest", this.serverRpc.LastRequestDispatched?.GetType().FullName);
+    }
+
+    [Fact]
+    public async Task DelegatedDispatcherCanDispatchInReverseOrder()
+    {
+        this.serverRpc.EnableBuffering = true;
+
+        var totalCallCount = 10;
+        var taskList = new List<Task<int>>();
+
+        for (int i = 0; i < totalCallCount; i++)
+        {
+            taskList.Add(this.clientRpc.InvokeAsync<int>(nameof(Server.GetCallCountAsync)));
+        }
+
+        await this.serverRpc.FlushRequestQueueAsync(totalCallCount);
+
+        for (int i = 0; i < totalCallCount; i++)
+        {
+            var result = await taskList[i];
+            Assert.Equal(totalCallCount - i, result);
+        }
+    }
+
+#pragma warning disable CA1801 // use all parameters
+    public class Server
+    {
+        private int callCounter = 0;
+
+        public Task TestMethodAsync()
+        {
+            return Task.CompletedTask;
+        }
+
+        public Task<int> GetCallCountAsync()
+        {
+            int currentCount = Interlocked.Increment(ref this.callCounter);
+            return Task.FromResult(currentCount);
+        }
+
+        public Task MethodThatThrowsAsync()
+        {
+            throw new InvalidProgramException();
+        }
+    }
+
+    public class DelegatedJsonRpc : JsonRpc
+    {
+        private List<(TaskCompletionSource<bool>, Task<JsonRpcMessage>)> requestSignalQueue = new List<(TaskCompletionSource<bool>, Task<JsonRpcMessage>)>();
+
+        public DelegatedJsonRpc(IJsonRpcMessageHandler handler)
+            : base(handler)
+        {
+        }
+
+        public DelegatedJsonRpc(IJsonRpcMessageHandler handler, object target)
+            : base(handler, target)
+        {
+        }
+
+        public bool EnableBuffering { get; set; }
+
+        public JsonRpcRequest? LastRequestDispatched { get; private set; }
+
+        public async Task FlushRequestQueueAsync(int expectedCount)
+        {
+            while (this.requestSignalQueue.Count != expectedCount)
```

Comment:
This loop looks like you're expecting another thread to enqueue messages. Yet you have no thread-safety protections in place.
What if you use `AsyncQueue<T>` instead, which is thread-safe and lets you simply `await DequeueAsync` `expectedCount` times in a loop?

Reply:
I will replace with AsyncQueue, The delay was enough 99.9% cases for queue to be all filled anyway in the test case but I agree there was still chance of failure.